src/main/kotlin/treeLib/bintrees/RBTree.kt

```diff
@@ -0,0 +1,204 @@
+                }
+                parent = treeBranch.first()
+                val temp = copyTo.value
+                copyTo.key = curNode.key
+                copyTo.value = curNode.value
+                curNode.value = temp  //to save deleted node's value
+                curNode.left
+            }
+            sonRight != null -> sonRight
+            else /*sonLeft != null*/ -> sonLeft
+        }
+        amountOfNodes--
+        when (curNode) {   //here we are "deleting" the node and replace it with his son
+            parent?.right -> parent.right = sonOfRemovedNode
+            parent?.left -> parent.left = sonOfRemovedNode
+            else -> root = sonOfRemovedNode   //only root doesn't have parent
+        }
+        treeBranch.addFirst(sonOfRemovedNode)
+        //curNode.isRed -> no balancing needed
+        if (!curNode.isRed) balancedRemove(treeBranch)
+        return curNode.value
+    }
+
+    /**
+     * Used in remove() method to balance tree :(
+     */
+    private fun balancedRemove(treeBranch: ArrayDeque<RBNode<K, V>?>) {
+        var son = treeBranch.removeFirstOrNull()
+        var parent = treeBranch.removeFirstOrNull()
+        var grandparent = treeBranch.removeFirstOrNull()
+        //son == son of removed node
+        if (son?.isRed == true) {
+            son.isRed = false
+            return
+        }
+        while (parent != null) {
+            var brother = if (son == parent.right) parent.left else parent.right
+            //we want for brother to be black (´。＿。｀)
+            if (brother?.isRed == true) {
+                brother.isRed = false
+                parent.isRed = true
+                if (son == parent.right) {
+                    rotateRight(parent, grandparent)
+                    grandparent = brother
+                    brother = parent.left
+                } else /*son == parent.left*/ {
+                    rotateLeft(parent, grandparent)
+                    grandparent = brother
+                    brother = parent.right
+                }
+            }
+            brother ?: throw IllegalStateException(
+                "Balance of red-black tree violated:" +
+                    " parent node have subtrees with black height 2 and black height 1"
+            )
+            //depending on nephew's colors, doing magic for balancing purposes
+            var nephewRight = brother.right
+            var nephewLeft = brother.left
+            if ((nephewRight == null || !nephewRight.isRed) && (nephewLeft == null || !nephewLeft.isRed)) {
+                brother.isRed = true
+                if (!parent.isRed) {   //parent is black --> go higher up
+                    son = parent       //the treeBranch and balance again
+                    parent = grandparent
+                    grandparent = treeBranch.removeFirstOrNull()
+                } else {
+                    parent.isRed = false
+                    break
+                }
+            } else {
```

Comment:
В код в первом if, если и выполняет, то после ничего не идет, поэтому мы можем избавиться от вложенности

```suggestion
            if ((nephewRight == null || !nephewRight.isRed) && (nephewLeft == null || !nephewLeft.isRed)) {
                brother.isRed = true
                if (!parent.isRed) {   //parent is black --> go higher up
                    son = parent       //the treeBranch and balance again
                    parent = grandparent
                    grandparent = treeBranch.removeFirstOrNull()
                } else {
                    parent.isRed = false
                    break
                }
                continue
            }
```

Reply:
Разделение логических блоков новой строкой очень увеличивает читабельность кода, можешь сравнить:

До:
```kotlin
    private fun balancedRemove(treeBranch: ArrayDeque<RBNode<K, V>?>) {
        var son = treeBranch.removeFirstOrNull()
        var parent = treeBranch.removeFirstOrNull()
        var grandparent = treeBranch.removeFirstOrNull()
        //son == son of removed node
        if (son?.isRed == true) {
            son.isRed = false
            return
        }
        while (parent != null) {
            var brother = if (son == parent.right) parent.left else parent.right
            //we want for brother to be black (´。＿。｀)
            if (brother?.isRed == true) {
                brother.isRed = false
                parent.isRed = true
                if (son == parent.right) {
                    rotateRight(parent, grandparent)
                    grandparent = brother
                    brother = parent.left
                } else /*son == parent.left*/ {
                    rotateLeft(parent, grandparent)
                    grandparent = brother
                    brother = parent.right
                }
            }
            brother ?: throw IllegalStateException(
                "Balance of red-black tree violated:" +
                    " parent node have subtrees with black height 2 and black height 1"
            )
            //depending on nephew's colors, doing magic for balancing purposes
            var nephewRight = brother.right
            var nephewLeft = brother.left
            if ((nephewRight == null || !nephewRight.isRed) && (nephewLeft == null || !nephewLeft.isRed)) {
                brother.isRed = true
                if (!parent.isRed) {   //parent is black --> go higher up
                    son = parent       //the treeBranch and balance again
                    parent = grandparent
                    grandparent = treeBranch.removeFirstOrNull()
                } else {
                    parent.isRed = false
                    break
                }
            } else {
                if (son == parent.right) {
                    if (nephewRight?.isRed == true && (nephewLeft == null || !nephewLeft.isRed)) {
                        rotateLeft(brother, parent)   //      B           /nL == new brother
                        nephewLeft = brother          //     / \  ->     /B   == new left nephew
                        brother = nephewRight         //    nL  nR      /nR
                        brother.isRed = false
                        nephewLeft.isRed = true
                    }
                    rotateRight(parent, grandparent)
                    brother.isRed = parent.isRed
                    parent.isRed = false
                    nephewLeft?.isRed = false
                } else /*son == parent.left*/ {
                    if (nephewLeft?.isRed == true && (nephewRight == null || !nephewRight.isRed)) {
                        rotateRight(brother, parent)  //      B       \nR == new brother
                        nephewRight = brother         //     / \  ->   \B == new right nephew
                        brother = nephewLeft          //    nL  nR      \nL
                        brother.isRed = false
                        nephewRight.isRed = true
                    }
                    rotateLeft(parent, grandparent)
                    brother.isRed = parent.isRed
                    parent.isRed = false
                    nephewRight?.isRed = false
                }
                break
            }
        }
        root?.isRed = false
    }
```

После:
```kotlin
    private fun balancedRemove(treeBranch: ArrayDeque<RBNode<K, V>?>) {
        var son = treeBranch.removeFirstOrNull()
        var parent = treeBranch.removeFirstOrNull()
        var grandparent = treeBranch.removeFirstOrNull()
        
        //son == son of removed node
        if (son?.isRed == true) {
            son.isRed = false
            return
        }

        while (parent != null) {
            var brother = if (son == parent.right) parent.left else parent.right

            //we want for brother to be black (´。＿。｀)
            if (brother?.isRed == true) {
                brother.isRed = false
                parent.isRed = true

                if (son == parent.right) {
                    rotateRight(parent, grandparent)
                    grandparent = brother
                    brother = parent.left
                } else /*son == parent.left*/ {
                    rotateLeft(parent, grandparent)
                    grandparent = brother
                    brother = parent.right
                }
            }

            brother ?: throw IllegalStateException(
                "Balance of red-black tree violated:" +
                        " parent node have subtrees with black height 2 and black height 1"
            )

            //depending on nephew's colors, doing magic for balancing purposes
            var nephewRight = brother.right
            var nephewLeft = brother.left
            if (isBlack(nephewRight) && isBlack(nephewLeft)) {
                brother.isRed = true

                if (!parent.isRed) {   //parent is black --> go higher up
                    son = parent       //the treeBranch and balance again
                    parent = grandparent
                    grandparent = treeBranch.removeFirstOrNull()
                } else {
                    parent.isRed = false
                    break
                }

                continue
            }

            if (son == parent.right) {
                if (nephewRight?.isRed == true && isBlack(nephewLeft)) {
                    rotateLeft(brother, parent)   //      B           /nL == new brother
                    nephewLeft = brother          //     / \  ->     /B   == new left nephew
                    brother = nephewRight         //    nL  nR      /nR
                    brother.isRed = false
                    nephewLeft.isRed = true
                }

                rotateRight(parent, grandparent)
                brother.isRed = parent.isRed
                parent.isRed = false
                nephewLeft?.isRed = false
            } else /*son == parent.left*/ {
                if (nephewLeft?.isRed == true && isBlack(nephewRight)) {
                    rotateRight(brother, parent)  //      B       \nR == new brother
                    nephewRight = brother         //     / \  ->   \B == new right nephew
                    brother = nephewLeft          //    nL  nR      \nL
                    brother.isRed = false
                    nephewRight.isRed = true
                }

                rotateLeft(parent, grandparent)
                brother.isRed = parent.isRed
                parent.isRed = false
                nephewRight?.isRed = false
            }

            break
        }

        root?.isRed = false
    }
```